src/JuMP.jl

```diff
@@ -53,38 +53,92 @@ include("utils.jl")
 const _MOIVAR = MOI.VariableIndex
 const _MOICON{F,S} = MOI.ConstraintIndex{F,S}
 
-# OptimizerFactory was deprecated in JuMP 0.21 and should be removed when
-# with_optimizer is removed.
-struct OptimizerFactory
-    # The constructor can be
-    # * `Function`: a function, or
-    # * `DataType`: a type, or
-    # * `UnionAll`: a type with missing parameters.
-    constructor
-    args::Tuple
-    kwargs # type changes from Julia v0.6 to v0.7 so we leave it untyped for now
-end
-
-function with_optimizer(constructor,
-                        args...; kwargs...)
-    deprecation_message = """
-with_optimizer is deprecated. The examples below demonstrate how to update to the new syntax:
-- 'with_optimizer(Ipopt.Optimizer)' becomes 'Ipopt.Optimizer'.
-- 'set_optimizer(model, with_optimizer(Ipopt.Optimizer, print_level=1, tol=1e-5))' becomes 'set_optimizer(model, Ipopt.Optimizer); set_parameters(model, \"print_level\" => 1, \"tol\" => 1e-5)'.
-- In rare cases where an argument must be passed to the constructor, use an anonymous function. For example, 'env = Gurobi.Env(); set_optimizer(model, with_optimizer(Gurobi.Optimizer, env))' becomes 'env = Gurobi.Env(); set_optimizer(model, () -> Gurobi.Optimizer(env))'.
-    """
-    Base.depwarn(deprecation_message, :with_optimizer)
-    if !applicable(constructor, args...)
-        error("$constructor does not have any method with arguments $args.",
-              "The first argument of `with_optimizer` should be callable with",
-              " the other argument of `with_optimizer`.")
-    end
-    return OptimizerFactory(constructor, args, kwargs)
+"""
+    optimizer_with_attributes(optimizer_constructor, params::Pair...)
+
+Groups an optimizer constructor with the list of parameters `params`. Note that
+it is equivalent to `MOI.OptimizerWithAttributes`.
+
+## Examples
+
+The call `optimizer_with_attributes(Gurobi.Optimizer, "Presolve" => 0)` groups
+a Gurobi optimizer with the parameter `Presolve` with value `0`. When
+provided to the `Model` constructor or to [`set_optimizer`](@ref), it creates a
+Gurobi optimizer and then set the `Presolve` parameter to value `0`:
+```julia
+julia> model = Model(optimizer_with_attributes(Gurobi.Optimizer, "Presolve" => 0));
+
+julia> MOI.get(model, MOI.RawParameter("Presolve"))
+0
+```
+Alternatively, the parameter can be provided separately with either
+[`set_parameter`](@ref):
+```julia
+julia> model = Model(Gurobi.Optimizer);
+
+julia> set_parameter(model, "Presolve", 0)
+0
+
+julia> MOI.get(model, MOI.RawParameter("Presolve"))
```

Comment:
I guess we should have `get_parameter`?

Also, did we make a decision about deprecating `set_parameter` in favor of `set_attribute`?

Reply:
Not sure, we should open an issue to discuss this

Reply:
See https://github.com/JuliaOpt/JuMP.jl/issues/2149